### PR TITLE
406 Error for application and Resource

### PR DESCRIPTION
**Pull request by dscervantes**
## What?
Explain what the change is linking any relevant JIRAs or Issues.
Adding 406 functionality 

## Why?
Consider what business or engineering goal does this PR achieves.
When an application or a resource does not match it does not throw an Error it just leaves the payload to pending then deleted payload after 7 days

## How?
Describe how the change is implemented. Any noteable new libaries, APIs, or features.
add new error 

## Testing
Did you add any tests for the change?

## Anything Else?
Any other notes about the PR that would be useful for the reviewer. 

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [X] General Coding Practices

### CODE REVIEW

**Review by dehort — CHANGES_REQUESTED** [review]
Looking good.  I left a couple of comments.

Also, please move that chunk of logic into a function.  This way the core logic of the PostExport method is simplified a bit.

File: exports/exports.go [hunk 1]
```diff
@@ -68,6 +69,33 @@ func (e *Export) PostExport(w http.ResponseWriter, r *http.Request) {
...
+		for _, resource := range cfg.ExportableApplications[source.Application]{
+			if resource == source.Resource {
+				resourceFound = true
+			}
+		}
+		if !resourceFound {
+			logger.Errorw("invalid resource","error",err)
+			StatusNotAcceptableError(w, "wrong resource")
+			return
+		}
```
[the remark this answers is not in the document]
Comment (dehort) [reply exports/exports.go:96]: @dscervantes Here is an example of how you could convert a `map[string][]string` to a `map[string]map[string]bool` and check to see if an application/resource combination is "exportable":  https://go.dev/play/p/iGjO7TabhHT

Let me know if you have any questions

[the remark this answers is not in the document]
Comment (dscervantes) [reply exports/exports.go]: I believed i made the appropriate changes let me know!

**Comment by dscervantes** [comment]
@dehort whenever you have time!

File: config/config.go [hunk 2]
```diff
@@ -128,16 +130,17 @@ func Get() *ExportConfig {
...
+			PublicPort:         	options.GetInt("PUBLIC_PORT"),
+			MetricsPort:        	options.GetInt("METRICS_PORT"),
+			PrivatePort:        	options.GetInt("PRIVATE_PORT"),
+			Debug:              	options.GetBool("DEBUG"),
+			LogLevel:           	options.GetString("LOG_LEVEL"),
+			OpenAPIPublicPath:  	options.GetString("OPEN_API_FILE_PATH"),
+			OpenAPIPrivatePath: 	options.GetString("OPEN_API_PRIVATE_PATH"),
+			Psks:               	options.GetStringSlice("PSKS"),
+			ExportExpiryDays:   	options.GetInt("EXPORT_EXPIRY_DAYS"),
+			ExportableApplications: options.GetStringMapStringSlice("EXPORT_ENABLE_APPS"),
```
Comment (dehort) [inline config/config.go:143]: I would perform the conversion from `map[string][]string` to `map[string]map[string]bool` right here.  That way we just have to do it one time and then the code in the `exports` module just has to check it.

Something like this should work:

```
var exportableAppsFromConfig map[string][]string
ExportableApplications: convertExportableAppsFromConfigToInternal(options.GetStringMapStringSlice("EXPORT_ENABLE_APPS")),
```

File: config/config.go [hunk 3]
```diff
@@ -270,3 +273,19 @@ func buildBaseHttpUrl(tlsEnabled bool, hostname string, port int) string {
 
 	return fmt.Sprintf("%s://%s:%d", protocol, hostname, port)
 }
+
+func ConvertConfigtoInternal(config map[string][]string) map[string]map[string]bool {
```
Comment (dehort) [inline config/config.go:277]: I would make this method private to the module.  With go lang, "private" functions begin with a lower case letter.

File: exports/exports.go [hunk 4]
```diff
@@ -68,17 +69,32 @@ func (e *Export) PostExport(w http.ResponseWriter, r *http.Request) {
...
+	if len(apiExport.Sources) == 0 {
+		logger.Errorw("no sources provided", "error", err)
+		BadRequestError(w, "no sources provided")
+		return
+	}
+
+	cfg := config.Get()
+
+	var exportableApplications = config.ConvertConfigtoInternal(cfg.ExportableApplications)
+
```
Comment (dehort) [inline exports/exports.go:81]: The conversion could be moved into the `config` module as described above.

File: exports/exports.go [hunk 5]
```diff
@@ -68,17 +69,32 @@ func (e *Export) PostExport(w http.ResponseWriter, r *http.Request) {
...
+		logger.Errorw("no sources provided", "error", err)
+		BadRequestError(w, "no sources provided")
+		return
+	}
+
+	cfg := config.Get()
+
+	var exportableApplications = config.ConvertConfigtoInternal(cfg.ExportableApplications)
+	
+	for _, source:= range apiExport.Sources{
```
Comment (dehort) [inline exports/exports.go:82]: Moving the `for` loop into the `verifyExportableApplication()` would make this a bit easier to read.

**Review by dehort — CHANGES_REQUESTED** [review]
Looking good.  I left a couple of comments.

**Comment by dehort** [comment]
@dscervantes It would be good to add a test to verify that an "invalid" app/resource returns a 406/error.

**Comment by dscervantes** [comment]
Added test let me know if I need to change it

File: exports/exports_test.go [hunk 6]
```diff
@@ -537,6 +537,23 @@ var _ = Describe("The public API", func() {
...
+			"Test Export Request",
+			"json",
+			"",
+			`{"application":"exampleApp", "resource":"exampleResource"}`,
+		)
+		AddDebugUserIdentity(req)
+		router.ServeHTTP(rr, req)
+		Expect(rr.Code).To(Equal(http.StatusNotAcceptable))
+		Expect(rr.Body.String()).To(ContainSubstring("Payload does not match Configured Exports"))
+	})
```
Comment (dehort) [inline exports/exports_test.go:556]: This test looks good!

It would good to add another version of this test that uses a valid app, but an invalid resource.  You could modify this test to be a table driven test like this:  https://github.com/RedHatInsights/export-service-go/actions/runs/9764633574/job/26953420583?pr=150

The syntax is a bit hard to follow, so reach out if you need a hand!

**Review by dehort — CHANGES_REQUESTED** [review]
Looking good!  I added a couple more comments.

**Comment by dscervantes** [comment]
@dehort  @skarekrow I believe my test are making the iqe plugin, any guidance there?

**Review by dehort — CHANGES_REQUESTED** [review]
This looks great!  However, it looks like the ginkgo tests are failing, so we need to get that fixed.

**Comment by dehort** [comment]
Hmm...the tests pass for me when running them locally  :confused:

**Comment by dehort** [comment]
@dscervantes can you try adding these lines to the clowdapp.yml file?

```
diff --git deploy/clowdapp.yaml deploy/clowdapp.yaml
index 15308d1..c2c273a 100644
--- deploy/clowdapp.yaml
+++ deploy/clowdapp.yaml
@@ -74,6 +74,8 @@ objects:
             secretKeyRef:
               name: export-service-psks
               key: psk-list
+        - name: EXPORT_ENABLE_APPS
+          value: ${EXPORT_ENABLE_APPS}
         initContainers:
         - args:
           - export-service
```

This should set the `EXPORT_ENABLE_APPS` environment variable  for the service when its deployed via clowder.

**Comment by dehort** [comment]
@dscervantes can you try this? 

```
diff --git cmd/export-service/api_server.go cmd/export-service/api_server.go
index 2d6cdd0..0b974c7 100644
--- cmd/export-service/api_server.go
+++ cmd/export-service/api_server.go
@@ -167,6 +167,7 @@ func startApiServer(cfg *config.ExportConfig, log *zap.SugaredLogger) {
                "debug", cfg.Debug,
                "publicopenapifilepath", cfg.OpenAPIPublicPath,
                "privateopenapifilepath", cfg.OpenAPIPrivatePath,
+               "exportableApplications", cfg.ExportableApplications,
        )
 
        kafkaProducerMessagesChan := make(chan *kafka.Message) // TODO: determine an appropriate buffer (if one is actually necessary)
diff --git deploy/clowdapp.yaml deploy/clowdapp.yaml
index c2c273a..d8b8db6 100644
--- deploy/clowdapp.yaml
+++ deploy/clowdapp.yaml
@@ -199,4 +199,4 @@ parameters:
   - name: LOG_LEVEL
     value: INFO
   - name: EXPORT_ENABLE_APPS
-    value: "{\"exampleApp\":[\"exampleResource\", \"anotherExampleResource\"]}"
+    value: "{\"exampleApplication\":[\"exampleResource\", \"anotherExampleResource\"]}"

```

**Comment by dehort** [comment]
The tests appear to work with those changes here:  https://github.com/RedHatInsights/export-service-go/pull/162

**Review by dehort — APPROVED** [review]
Looks good to me!
